### PR TITLE
fix: Theming UX Changes

### LINKED ISSUE #13629: [Feature]: Improve the primary and background color representation in app theming
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Today in the theming pane, we have a color pane with 2 colors but there is some confusion if these colors can be edited or not because of the difference in UI between these colors and the remaining ones. 

Screenshot 2022-05-06 at 10 14 06 AM
Screenshot 2022-05-06 at 10 14 20 AM

### Why should this be worked on?

Consistency with other color picker UIs

### PR DESCRIPTION
**Pull request by jsartisan**
## Description
The UX for color controls in theming Editor was little confusing. This PR updates the UX with changes suggested in the issue.

Fixes #13629

> if no issue exists, please create an issue and ask the maintainers about this first

## Type of chan

- Bug fix

## How Has This Been Tested?

- manually tested

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: theming/next-steps 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.74 **(0.03)** | 38.68 **(0.04)** | 36.27 **(0.04)** | 56.99 **(0.04)**
 :red_circle: | app/client/src/actions/appThemingActions.tsx | 67.65 **(-0.09)** | 100 **(0)** | 0 **(0)** | 70.59 **(-2.74)**
 :green_circle: | app/client/src/components/ads/ColorPickerComponentV2.tsx | 79.38 **(0.53)** | 72.41 **(0.27)** | 75.76 **(0.76)** | 79.19 **(0.42)**
 :sparkles: :new: | **app/client/src/entities/AppTheming/utils.ts** | **85.48** | **60.71** | **75** | **83.64**
 :red_circle: | app/client/src/pages/Editor/ThemePropertyPane/ThemeEditor.tsx | 77.55 **(-0.71)** | 100 **(0)** | 11.11 **(-1.39)** | 76.6 **(-0.67)**
 :green_circle: | app/client/src/pages/Editor/ThemePropertyPane/controls/ThemeColorControl.tsx | 86.96 **(-2.51)** | 66.67 **(16.67)** | 60 **(0)** | 83.33 **(-4.17)**
 :red_circle: | app/client/src/sagas/AppThemingSaga.tsx | 31.82 **(-0.23)** | 6.67 **(-1.02)** | 9.09 **(-0.91)** | 36.36 **(0.13)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**
 :green_circle: | app/client/src/widgets/WidgetUtils.ts | 85.42 **(2.61)** | 65.22 **(3.48)** | 73.33 **(6.66)** | 82.28 **(3.17)**
 :sparkles: :new: | **app/client/src/widgets/TableWidget/widget/helpers.ts** | **42.86** | **100** | **0** | **33.33**
 :green_circle: | app/client/src/widgets/TableWidget/widget/propertyConfig.ts | 95.56 **(0.11)** | 100 **(0)** | 92.31 **(0)** | 95.56 **(0.11)**</details>

### CODE REVIEW

File: app/client/src/sagas/AppThemingSaga.tsx [hunk 1]
```diff
@@ -253,10 +266,148 @@ function* closeisBetaCardShown() {
...
+
+                Object.keys(fieldStylesheet).map((fieldPropertyKey) => {
+                  const fieldStylesheetValue =
+                    fieldStylesheet[fieldPropertyKey];
+
+                  if (
+                    fieldStylesheetValue !== get(schemaItem, fieldPropertyKey)
+                  ) {
+                    modifications[
+                      `${[propertyPath]}.${fieldPropertyKey}`
```
Comment (ashit-rath) [inline app/client/src/sagas/AppThemingSaga.tsx:364]: Check if you are able to access and use `formData`/`sourceData`/`fieldState` in the JSONForm's field's `boxShadow` `borderRadius` property and there isn't any odd behaviour when you toggle the JS mode. If there is an issue then you might have to do the same thing that you did for the table widget.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/AppThemingTests/Application_Theme_MultiSelectWidget_spec.js [hunk 2]
```diff
@@ -105,16 +105,14 @@ describe("Theme validation usecase for multi-select widget", function() {
```
Comment (ashit-rath) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/AppThemingTests/Application_Theme_MultiSelectWidget_spec.js:105]: Question: How are we testing the reset flow?

File: app/client/src/sagas/AppThemingSaga.tsx [hunk 3]
```diff
@@ -253,10 +266,148 @@ function* closeisBetaCardShown() {
   } catch (error) {}
 }
 
+/**
+ * resets widget styles
+ */
+function* resetTheme() {
```
Comment (ashit-rath) [inline app/client/src/sagas/AppThemingSaga.tsx:272]: We should add jest tests here.
Either you can move out the operations to helper functions and test them in isolation
or
you can test the saga with a bigger test data.
Or if you have a better strategy to test this then please do implement it.

**Review by ashit-rath — COMMENTED** [review]
PR title :-
1. Modify the title from `fix` to `feature`/`feat`
2. Make the title more descriptive so that it helps when release notes are created.

PR description: -
1. Please make sure the PR description checklist is checked correctly
2. Add some more description for this PR

File: app/client/src/pages/Editor/ThemePropertyPane/controls/ThemeColorControl.tsx [hunk 4]
```diff
@@ -20,8 +20,11 @@ const ColorBox = styled.div<{
 
 function ThemeColorControl(props: ThemeColorControlProps) {
   const { theme, updateTheme } = props;
-  const [selectedColor, setSelectedColor] = useState<string | null>(null);
+  const [autoFocus, setAutoFocus] = useState(false);
   const userDefinedColors = theme.properties.colors;
+  const [selectedColor, setSelectedColor] = useState<string | null>(
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/src/pages/Editor/ThemePropertyPane/controls/ThemeColorControl.tsx:25]: I think this was related to old ux when we were not selecting color by default. removing it.

File: app/client/src/sagas/AppThemingSaga.tsx [hunk 5]
```diff
@@ -253,10 +266,148 @@ function* closeisBetaCardShown() {
...
+    const canvasWidgets = yield select(getCanvasWidgets);
+    const themeStylesheet = yield select(getSelectedAppThemeStylesheet);
+
+    const propertiesToIgnore = [
+      "childStylesheet",
+      "submitButtonStyles",
+      "resetButtonStyles",
+    ];
+
+    // return;
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/src/sagas/AppThemingSaga.tsx:284]: Removed.

**Comment by shadabbuchh** [comment]
@dilippitchika / @jsartisan with regards to https://app.zenhub.com/workspaces/app-viewers-pod-616922264a5ea000108d319d/issues/appsmithorg/appsmith/13596 I have the following question:

- What should be the behaviour for properties like:
   a. `Background Color`
   b. `Item Background Color`
And in general for widgets which are similar to `Container` type? Should these properties also reset to the current selected theme style?

**Comment by shadabbuchh** [comment]
@jsartisan with regards to https://github.com/appsmithorg/appsmith/issues/13596 below are the observations made:

- For all widgets in general:
If a user selects some `Box Shadow` other than `None`. Now when the user clicks on the `Reset widget styles`... for widgets where `"boxShadow": "none"`, the shadow gets reset to `None`. Ideally shouldn't it remain the way the user had selected it in the first place?

https://images.zenhubusercontent.com/275993885/dee3cc54-51ee-420c-90b1-bd9d24aedc11/screen_recording_2022_06_01_at_12_47_10_pm.mov

- For Table widget:
   a. For columns of type `Button`, `Icon Button`, `Menu Button` when we change properties such as `Button Color`, `Icon Color`, `Menu Color`, `Border Radius` (which are configured in theme config) we don't see the yellow dot on changing their values. (We see this happening in other widgets as well)
   b. The Icon color for column `Icon Button` does not reset to the theme value when the user clicks on `Reset widget styles`.

https://images.zenhubusercontent.com/275993885/28e43a9a-0865-4ae3-9311-b8ed4e139f0f/screen_recording_2022_06_01_at_12_50_01_pm.mov

- When a user manually resets the value of any property to the theme value the `yellow dot` & the `Reset Value` arrow still appear on the screen e.g. if in the `Text` widget I go and change the `Font Family` to a different value and then again select the value that is consistent with the current theme, the `yellow dot` & the `Reset Value` arrow should ideally vanish (as I have manually reset the value to the theme value) but these still appear, which can be confusing for the user. Should this be taken as part of this issue or should I create a new one for it?

https://images.zenhubusercontent.com/275993885/22affd18-7d20-48c8-a79b-a3ab35e81d20/screen_recording_2022_06_01_at_2_08_36_pm.mov

**Comment by jsartisan** [comment]
@shadabbuchh 

> For all widgets in general:
If a user selects some Box Shadow other than None. Now when the user clicks on the Reset widget styles... for widgets where "boxShadow": "none", the shadow gets reset to None. Ideally shouldn't it remain the way the user had selected it in the first place?

This is fixed.

> For Table widget:
a. For columns of type Button, Icon Button, Menu Button when we change properties such as Button Color, Icon Color, Menu Color, Border Radius (which are configured in theme config) we don't see the yellow dot on changing their values. (We see this happening in other widgets as well)

This will be fixed separately because there are some issues happening with reset flow on properties.

> b. The Icon color for column Icon Button does not reset to the theme value when the user clicks on Reset widget styles.
This is fixed but you will be able to test it after it goes into release because it's a backend config change.

> When a user manually resets the value of any property to the theme value the yellow dot & the Reset Value arrow still appear on the screen e.g. if in the Text widget I go and change the Font Family to a different value and then again select the value that is consistent with the current theme, the yellow dot & the Reset Value arrow should ideally vanish (as I have manually reset the value to the theme value) but these still appear, which can be confusing for the user. Should this be taken as part of this issue or should I create a new one for it?

This is not an issue actually. We will show the orange dot if the value is not a binding defined in the config. This is happening on border radius and box shadow properties as well.

**Comment by shadabbuchh** [comment]
@jsartisan 

> For all widgets in general:
If a user selects some Box Shadow other than None. Now when the user clicks on the Reset widget styles... for widgets where "boxShadow": "none", the shadow gets reset to None. Ideally shouldn't it remain the way the user had selected it in the first place?

I see this happening now as well. Although it is fixed for most of the widgets but the shadow is reset for buttons in the `JSON Form` widget.

https://www.loom.com/share/292b3b1777e147958f66e96d951c12cd

**Comment by jsartisan** [comment]
@shadabbuchh the above callout is fixed.

**Comment by shadabbuchh** [comment]
> @shadabbuchh the above callout is fixed.

@jsartisan I have validated the above callout and it is working as expected. 

But now I see a new issue happening... In the `JSON Form` itself if I change the properties (`Accent Color` or `Border Radius`) of the default fields: `Name`, `Date of Birth`, `Employee Id` and then go and click on `Reset Widget Styles`, properties are reset and the JS value is binded to the theme but I still see the yellow dot and `Reset Value` arrow above these properties. As discussed this will be tracked as a separate issue https://github.com/appsmithorg/appsmith/issues/14220.

Hence I am moving https://github.com/appsmithorg/appsmith/issues/13596 to `Done`. Thanks.

**Comment by jsartisan** [comment]
The issues with orange dot will be fixed separately because the current logic for the orange dot does not work properly in table widget and json form. Here are the two issues for it - https://github.com/appsmithorg/appsmith/issues/14220 and https://github.com/appsmithorg/appsmith/issues/14219

File: app/client/src/components/ads/ColorPickerComponentV2.tsx [hunk 6]
```diff
@@ -315,6 +317,10 @@ function ColorPickerComponent(props: ColorPickerProps) {
     [],
   );
 
+  useEffect(() => {
+    setIsOpen(isOpenProp);
+  }, [isOpenProp]);
```
Comment (ashit-rath) [inline app/client/src/components/ads/ColorPickerComponentV2.tsx:322]: `setIsOpen` should be in deps of useEffect.

File: app/client/src/pages/Editor/ThemePropertyPane/ThemeCard.tsx [hunk 7]
```diff
@@ -25,13 +25,14 @@ import { getCurrentApplicationId } from "selectors/editorSelectors";
...
  */
-interface ThemeCard {
+type ThemeCard = React.PropsWithChildren<{
   theme: AppTheme;
   isSelected?: boolean;
   className?: string;
   selectable?: boolean;
   deletable?: boolean;
-}
+  children?: React.ReactElement[] | React.ReactElement;
```
Comment (ashit-rath) [inline app/client/src/pages/Editor/ThemePropertyPane/ThemeCard.tsx:34]: You can remove the `children` now as you are using the `PropsWithChildren`

File: app/server/appsmith-server/src/main/resources/system-themes.json [hunk 8]
```diff
@@ -322,7 +322,7 @@
             "boxShadow": "none"
           },
           "iconButton": {
-            "menuColor": "{{appsmith.theme.colors.primaryColor}}",
+            "buttonColor": "{{appsmith.theme.colors.primaryColor}}",
```
Comment (ashit-rath) [inline app/server/appsmith-server/src/main/resources/system-themes.json:325]: Why are we changing this?

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/GitSync/GitSyncedApps_spec.js [hunk 9]
```diff
@@ -373,11 +377,10 @@ describe("Git sync apps", function() {
     cy.get(commonlocators.backToEditor).click();
     cy.wait(2000);
     // verfiy data binding on all pages in edit mode
-    /* cy.get(".bp3-input")
-      .first()
-      .should("have.value", "morpheus");
-    cy.get(".bp3-input")
-      .eq(1)
+    /* cy.get(".t--draggable-inputwidgetv2").first().find(".bp3-input").should("have.value", "morpheus");
```
Comment (ashit-rath) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/GitSync/GitSyncedApps_spec.js:380]: Should we get rid of all the commented code here?

**Review by ashit-rath — COMMENTED** [review]
@jsartisan Minor callouts

File: app/server/appsmith-server/src/main/resources/system-themes.json (hunk 8 above)
Comment (jsartisan) [reply app/server/appsmith-server/src/main/resources/system-themes.json:325]: This was a mistake in our config. iconButton actually uses buttonColor. But we used menuColor instead which was wrong.

Comment (ashit-rath) [reply app/server/appsmith-server/src/main/resources/system-themes.json:325]: Got it, so `menuColor` never got used and `buttonColor` always got undefined.

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**